src/utils/marketsAndNetworksConfig.ts

```diff
@@ -143,6 +144,134 @@ export const isFeatureEnabled = {
   permissions: (data: MarketDataType) => data.enabledFeatures?.permissions,
 };
 
+function sleep(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+/*
+// Unused for now, need to debug why batch providers are incompatible with tenderly rpcs for fork/testing usage
+class StaticJsonRpcBatchProvider extends ethersProviders.JsonRpcBatchProvider {
+  async detectNetwork(): Promise<Network> {
+    let network = this.network;
+    if (network == null) {
+      network = await super.detectNetwork();
+
+      if (!network) {
+        logger.throwError('no network detected', Logger.errors.UNKNOWN_ERROR, {});
+      }
+
+      // If still not set, set it
+      if (this._network == null) {
+        // A static network does not support "any"
+        defineReadOnly(this, '_network', network);
+
+        this.emit('network', network, null);
+      }
+    }
+    return network;
+  }
+} */
+
+/**
+ * The provider will rotate rpcs on error.
+ * If provider rotates away from the first RPC, rotate back after a set interval to prioritize using most reliable RPC.
+ * If provider rotates through all rpcs, delay to avoid spamming rpcs with requests.
+ */
+class RotationProvider extends ethersProviders.BaseProvider {
+  readonly providers: StaticJsonRpcProvider[];
+  private currentProviderIndex = 0;
+  private lastFullRotationTimestamp = 0;
+  private firstRotationTimestamp = 0;
+  // after completing a full rotation of the RotationProvider, delay to avoid spamming rpcs with requests
+  private ROTATION_DELAY = 5000;
+  // if we rotate away from first rpc, return back after this delay
+  private FALL_FORWARD_DELAY = 60000;
+
+  constructor(urls: string[], chainId: number) {
+    super(chainId);
+    this.providers = urls.map((url) => new StaticJsonRpcProvider(url, chainId));
+  }
+
+  /**
+   * If provider rotates through all rpcs, delay to avoid spamming rpcs with requests
+   */
+  async delayRotation() {
+    const now = new Date().getTime();
+    const diff = now - this.lastFullRotationTimestamp;
+    if (diff < this.ROTATION_DELAY) {
+      await sleep(this.ROTATION_DELAY - diff);
+    }
```

Comment:
On the first cycle through, we won't actually wait for the `ROTATION_DELAY` because `this.lastFullRotationTimestamp` doesn't get set until _after_ the delay (in `rotateUrl`)

I'm not sure I'm following what this logic is supposed to be doing, would it make more sense if we just did an `await sleep(ROTATION_DELAY)`?

Reply:
I updated this to always just wait the rotation delay time and removed the `lastFullRotationTimestamp` tracking